### ISSUE TITLE
Crossfilter 1.4.8 with more than 32 dimension, group all not removing correctly during filter the 32nd dimension

### ISSUE THREAD

**Issue by aarontjl** [crossfilter/crossfilter#158]
With crossfilter 1.4.8, i have more than 32 dimension, and group each dimension. 

The first time I call 32nd dimension's `filterExact` and then call the other dimension's group `all` function, the group value is correct. 

But when I remove the filter and call 32nd dimension's `filterExact` function again, I find that the lower dimension's group is not removing the value correctly, but the higher dimension's group removes the value correctly.

Here is my demo: [demo url](https://jsfiddle.net/5b1dtkov/8/)

Through debugging, I found that when the `onlyExcept` function of array.js is called for verification when the value is removed, it always returns false. As shown in the picture: 

![picture url](https://i.stack.imgur.com/yFAtC.png)

Am hoping someone can suggest places to inspect either my code or the crossfilter code to find the problem.

**Comment by gordonwoodhull** [comment]
Hi @aarontjl, thanks for the report and the fiddle!

Could you provide more detail about what is incorrect? Running your fiddle, I see the same result both times:

<img width="323" alt="image" src="https://user-images.githubusercontent.com/1366709/93221853-3ce59a00-f73c-11ea-8fd7-e59320af829e.png">

**Comment by gordonwoodhull** [comment]
Thanks for the report @aarontjl!

I had trouble reproducing this due to quirks in the Chrome debugger. Apparently it holds references to live objects, so if you log an object and then make a change to the object, the old log might reflect *later changes*.

For that reason, your "correct" example was displaying the incorrect result from later in the test. Very confusing.

Anyway, here is [a fork](https://jsfiddle.net/gordonwoodhull/wc1jqLfh/5/) of your notebook which displays the value for 'b' of all bins of the group, using

```js
console.log(groups.map(g => g.all()[1].value));
```

This avoids the problem by not printing any objects, just arrays of plain values. It looks like ***all*** groups below 32 are unfiltered:

```
[1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0]
```

Continuing further, just printing the values before the first filter causes the values to increment again:

```
[2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 1, 0, 0]
```

And repeating the filter, we can increment yet again:
```
[3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 1, 0, 0]
```

[Crazy increment fiddle](https://jsfiddle.net/gordonwoodhull/wc1jqLfh/11/).

**Comment by aarontjl** [comment]
Thank you very much for your reply @gordonwoodhull! 

I'm sorry for the confusion caused by my unclear description.As you said, this is exactly the problem I am currently experiencing.

**Comment by gordonwoodhull** [comment]
Nothing was unclear, except the Chrome debugger! Weird behavior there.

Sorry I don't have an immediate fix or workaround for you. If you want to try to fix this yourself, I would be glad to help.

The first step would be to turn your fiddle into a test under `test\` and run `npm test` to watch it fail. 

I am going on vacation soon, so I may be slow to reply for the rest of the month.

**Comment by gordonwoodhull** [comment]
It's probably obvious, but the backstory is that >32 dimensions is one of the newer features contributed by the community since we forked from square/crossfilter.

As such, it is more likely to have bugs than the original features. Still, I'm surprised there is such a serious issue. Perhaps no one has applied a filter to a dimension over 31 before; they only had passive charts that drew top dimensions without allowing brushing.

**Comment by aarontjl** [comment]
Hello, @gordonwoodhull 

This bug seems to comes from the  onlyExcept function in array.js.

When I filter the 32nd dimension, all the dimensions before the 32nd will remove the filter items as follows:

The value of the removed item in the array is 2147483648(2^31) and the mask assigned to 2147483648.As shown below:
![picture url](https://i.stack.imgur.com/A1j2H.png)

After  `mask &= zero`, mask becomes -2147483648 but onlyOne is 2147483648, so return false.As shown below:
![picture url](https://i.stack.imgur.com/860JH.png)

lt seems that 2147483648 AND any negative number, the result is -2147483648, eventually produced this error. 

Finally, I tried to use `mask = (mask & zero) >>> 0` instead of `mask &= zero`, and everything seems to be working fine.

**Comment by gordonwoodhull** [comment]
Hi @aarontjl, thanks for the analysis!

I agree, this ambiguity between signed and unsigned numbers in JavaScript is a likely cause.

Would you be interested in contributing a PR to fix the issue? As I mentioned above, we just need a test case that fails currently, and the fix you suggest, preferably in separate commits.

**Comment by aarontjl** [comment]
Thanks for the reply！

I will try to provide a PR as soon as possible to fix the issue.

**Comment by gordonwoodhull** [comment]
Fixed in 1.5.4. Thanks @aarontjl!
